Optimization: Reduce the set of predicates by replacing subexpressions

wangandi opened the pull request:
**Updated description:**
Basically, what this PR does is reduce a certain predicates of the form `p1 & p2 & ... &pn` via subexpression matching. 

Example 1:
```
select * from t1 where f1 is null and f1 is not null;
```
Our code will break the predicate above into two components: `f1 is null` and `not(f1 is null)`. The PR code will see that `f1 is null` is a subexpression in `not(f1 is null)` and change it to `not(true)`. `not(true)` in turn reduces to `false`, and then the plan itself will ​reduce to an empty constant.

Example 2:
```
select * from t1 where f1 is not null and ((f1  = f2 + 1) or (f1 is null and f2 is null));
```
Our code will break the predicate above into two components: `not(f1 is null)` and `((f1  = f2 + 1) or (f1 is null and f2 is null))`. The PR code will see that `f1 is null` is a subexpression in `((f1  = f2 + 1) or (f1 is null and f2 is null))` and change it to `((f1  = f2 + 1) or (false and f2 is null))`. `((f1  = f2 + 1) or (false and f2 is null))` in turn reduces to `(f1  = f2 + 1)`, so the plan should look like: 
```
Get t1
Filter !(isnull(#0)), (#0 = (#1 + 1))
```

----------
**Limitations**

Because the sideways propagation is only limited to subexpression matching at the moment, something like what Philip tested in the comment below does not work:
```
materialize=> CREATE TABLE t1 (col_null INTEGER, col_not_null INTEGER NOT NULL);
materialize=> EXPLAIN SELECT * FROM t1 WHERE col_not_null = col_null AND col_null IS NULL;
           Optimized Plan           
------------------------------------
 %0 =                              +
 | Get materialize.public.t1 (u207)+
 | Filter isnull(#0), (#0 = #1)    +
```

Note that the subexpression matching in this PR requires an exact match, so something like this also does not work:
```
materialize=> CREATE TABLE t2 (col_null INTEGER NOT NULL, col_not_null INTEGER NOT NULL);
materialize=> EXPLAIN SELECT * FROM t2 WHERE col_not_null = col_null AND NOT col_not_null = col_null;
           Optimized Plan           
------------------------------------
%0 =
| Get materialize.public.t2 (u3)
| Filter (#0 = #1), (#0 != #1)
```
The reason is that our optimizer converts `not(#0 = #1)` into `(#0 neq #1)`, which does not match `(#0 eq #1)`. 

Stuff like `(#1 = #0)` might not successfully match `(#0 = #1)`.

To avoid correctness errors, the PR only does subexpression matching if a predicate is not nullable.

Compare this plan where the columns are all not nullable:
```
materialize=>  CREATE TABLE t2 (col_null INTEGER NOT NULL, col_not_null INTEGER NOT NULL);
materialize=> EXPLAIN SELECT * FROM t2 WHERE col_not_null = col_null AND ((col_not_null = col_null) or (col_not_null = 1));
           Optimized Plan           
----------------------------------
%0 =
| Get materialize.public.t2 (u3)
| Filter (#0 = #1)
```
To this plan:
```
materialize=> CREATE TABLE t3 (col_null INTEGER, col_not_null INTEGER);
materialize=> EXPLAIN SELECT * FROM t3 WHERE col_not_null = col_null AND ((col_not_null = col_null) or (col_not_null = 1));
           Optimized Plan           
----------------------------------
%0 =
| Get materialize.public.t3 (u5)
| Filter ((#0 = #1) || (#1 = 1)), (#0 = #1)
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/6196)
<!-- Reviewable:end -->

philip-stoev commented:
@wangandi can you provide a few more examples in which this optimization is supposed to kick in? I wrote several dozen queries with increasing complexity and I am only able to see the optimization in action with a query equivalent to the nested subquery from the joins.slt, where all columns are not even declared as ```NOT NULL```. 

For example, in this branch and in main the following very simple query does use propagation:

```
materialize=> CREATE TABLE t1 (col_null INTEGER, col_not_null INTEGER NOT NULL);
materialize=> EXPLAIN SELECT * FROM t1 WHERE col_not_null = col_null AND col_null IS NULL;
           Optimized Plan           
------------------------------------
 %0 =                              +
 | Get materialize.public.t1 (u207)+
 | Filter isnull(#0), (#0 = #1)    +
```

wangandi commented:
@philip-stoev I have added some example tests to the PR and updated the PR description.

Let me know if you have any other questions on how the PR works!

philip-stoev commented:
I think this is OK, just for the record, the following degenerate query is better optimized without this patch:

```
materialize=>  EXPLAIN  SELECT *  FROM t2  AS a1  WHERE  NOT f1  IN ( 0 , f2 , f2 , f2  , f2  ,  f2, f2 );
                                       Optimized Plan                                       
--------------------------------------------------------------------------------------------
 %0 =                                                                                      +
 | Get materialize.public.t2 (u215)                                                        +
 | Filter (#0 != #1), (#0 != #1), (#0 != #1), (#0 != #1), (#0 != #1), (#0 != #1), (#0 != 0)+
```

The IN list is flattened better with the prior commit:

```
materialize=> EXPLAIN  SELECT *  FROM t2  AS a1  WHERE  NOT f1  IN ( 0 , f2 , f2 , f2  , f2  ,  f2, f2 );
          Optimized Plan          
----------------------------------
 %0 =                            +
 | Get materialize.public.t2 (u5)+
 | Filter (#0 != #1), (#0 != 0)  +
```

I do not think this is a very realistic case, so I am mentioning it here mostly for the record.

philip-stoev commented on src/expr/src/relation/canonicalize.rs:
```diff
@@ -83,6 +87,85 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+            |expr: &MirScalarExpr, constant_bool: &MirScalarExpr| {
+                // Do not replace subexpressions equal to `expr` if `expr` is a
+                // literal to avoid infinite looping.
+                if !expr.is_literal() && !expr.typ(input_type).nullable {
```
I am not sure if the code has the concept of a ```deterministic``` function but I think non-deterministic expressions should also be excluded from the optimization.

Right now there seems to be only one such function, ```mz_sleep``` and expressions containing it are reduced aggressively:

```
materialize=> EXPLAIN SELECT * FROM t1 WHERE mz_internal.mz_sleep(col_not_null) > mz_internal.mz_sleep(col_not_null) AND (mz_internal.mz_sleep(col_not_null) > mz_internal.mz_sleep(col_not_null) = true);
                       Optimized Plan                       
------------------------------------------------------------
 %0 =                                                      +
 | Get materialize.public.t1 (u227)                        +
 | Filter (mz_sleep(i32tof64(#1)) > mz_sleep(i32tof64(#1)))+
```

So the 4 sleeps have been reduced to 2. Without delving into @frankmcsherry's question as to whether in SQL one is required to evaluate the entire expression or only parts thereof, it seems safer to simply exclude non-deterministic expressions from any such optimizations/rewrites.

philip-stoev commented on test/sqllogictest/transform/predicate_reduction.slt:
```diff
@@ -0,0 +1,96 @@
+1
+0
+
+# A test that simplification works when overlapping subexpressions are nested.
```
This query is correctly simplified:

```
CREATE TABLE t1 (col_null INTEGER, col_not_null INTEGER NOT NULL);

materialize=> EXPLAIN SELECT * FROM t1 WHERE (col_not_null + 2) = 1 AND (((col_not_null + 2) = 1) = true);
           Optimized Plan           
------------------------------------
 %0 =                              +
 | Get materialize.public.t1 (u233)+
 | Filter (1 = (#1 + 2))           +
```

But this one, which differs only in the arithmetic operator used, is not:

```
materialize=> EXPLAIN SELECT * FROM t1 WHERE (col_not_null % 2) = 1 AND (((col_not_null % 2) = 1) = true);
                  Optimized Plan                  
--------------------------------------------------
 %0 =                                            +
 | Get materialize.public.t1 (u233)              +
 | Filter (true = (1 = (#1 % 2))), (1 = (#1 % 2))+
```

wangandi replied:
> But this one, which differs only in the arithmetic operator used, is not:
> 
> ```
> materialize=> EXPLAIN SELECT * FROM t1 WHERE (col_not_null % 2) = 1 AND (((col_not_null % 2) = 1) = true);
>                   Optimized Plan                  
> --------------------------------------------------
>  %0 =                                            +
>  | Get materialize.public.t1 (u233)              +
>  | Filter (true = (1 = (#1 % 2))), (1 = (#1 % 2))+
> ```

I looked at the code determining nullability, and it seems that the cause of the lack of simplification is as follows:
We used to return null when there is an evaluation error. So, `select * from t1.f1/t1.f2 > 0` would return null if any row had `t1.f2 = 0`. We have since changed the behavior so we return to an evaluation error instead, but we forgot to change the code determining whether the predicate can return null in light of this.
We were also very lazy in determining if a predicate can be null, so if the function is `%`, then we automatically return that the predicate can be null even if the second argument is a non-zero literal.

philip-stoev commented on test/testdrive/subexpression-replacement.td:
```diff
@@ -0,0 +1,120 @@
+
+# NULL-able expressions are not optimized
+> EXPLAIN SELECT * FROM t1 WHERE col_null = 1 AND (col_null = 1 AND col_not_null = 5);
+"%0 =| Get t1| Filter (#0 = 1), (#0 = 1), (#1 = 5)"
```
I believe this is not optimized because of the fact that the expression is NULL-able, even though this particular example appears save to optimize.

wangandi replied:
Ah, this was because I took out the predicate dedupping code, thinking that subexpression matching replaced it, forgetting that subexpression matching doesn't run when the expression is nullable. Thanks for catching this!

philip-stoev commented on test/testdrive/subexpression-replacement.td:
```diff
@@ -0,0 +1,120 @@
+# Nested OR/disjunctions are not optimized
+
+> EXPLAIN SELECT * FROM t1 WHERE col_null IS NULL OR (col_null IS NULL OR col_not_null = 5);
+"%0 =| Get t1| Filter (isnull(#0) || (isnull(#0) || (#1 = 5)))"
```
This is not optimized I believe because nested-ORs are not flattened. I will open a separate ticket about flattening the ORs.

philip-stoev commented on test/testdrive/subexpression-replacement.td:
```diff
@@ -0,0 +1,120 @@
+# A more complex expression
+
+> EXPLAIN SELECT * FROM t1 WHERE (col_not_null + 1 / col_not_null) = 5 AND ((col_not_null + 1 / col_not_null) = 5 AND col_null = 6);
+"%0 =| Get t1| Filter (#0 = 6), (5 = (#1 + (1 / #1))), (5 = (#1 + (1 / #1)))"
```
This could have been optimized, but was not.

philip-stoev commented on test/testdrive/subexpression-replacement.td:
```diff
@@ -0,0 +1,120 @@
+
+# This is not optimized
+> EXPLAIN SELECT * FROM t1 WHERE (col_not_null % 2) = 1 AND (((col_not_null % 2) = 1) = TRUE);
+"%0 =| Get t1| Filter (true = (1 = (#1 % 2))), (1 = (#1 % 2))"
```
This is the ```%``` example mentioned elsewhere in the review.

philip-stoev commented on test/testdrive/subexpression-replacement.td:
```diff
@@ -0,0 +1,120 @@
+> EXPLAIN SELECT * FROM t1
+  WHERE mz_internal.mz_sleep(col_not_null) > mz_internal.mz_sleep(col_not_null)
+  AND (mz_internal.mz_sleep(col_not_null) > mz_internal.mz_sleep(col_not_null) = true);
+"%0 =| Get t1| Filter (mz_sleep(i32tof64(#1)) > mz_sleep(i32tof64(#1)))"
```
This is the deterministic function thing discussed elsewhere.

philip-stoev commented on test/testdrive/subexpression-replacement.td:
```diff
@@ -0,0 +1,120 @@
+# Not optimized in OR/disjunctions
+
+> EXPLAIN SELECT * FROM t1 WHERE col_not_null = 1 OR TRUE IN (col_not_null = 1, col_not_null = 2)
+"%0 =| Get t1| Filter ((#1 = 1) || ((true = (#1 = 1)) || (true = (#1 = 2))))"
```
The fact that this is not optimized is I believe due to the fact that the IN-list produces a nested OR .

frankmcsherry commented:
> Right now there seems to be only one such function, `mz_sleep` and expressions containing it are reduced aggressively:

This was one of the warnings wtih `mz_sleep`, that the optimizer might do weird things to it. I don't think this needs to turn in to optimizer debt to sort out.

philip-stoev reviewed (approved):
With a randomized workload:
- I am able to trigger the optimization and it seems to have a net positive effect for the vast majority of the queries
- no wrong results

With the manual tests that I wrote:
- I have added review comments for all the queries from the test that are not being optimized
- The .td test is my first attempt to have many query plans inside a .td test and I will be using it to hone testdriver so that one can manipulate such tests faster in the future (e.g. I hope to have testdrive stop failing at the first query  but rather be able to see all changed/different plans together, be able to modify such tests in builk account fo any changed plans, etc.). For the time  being, this test just used a regular expression to make the query plans more concise.

frankmcsherry commented:
I'm not sure that this optimization is correct, and it would be great to discuss to determine if it is. If I have two predicates, let's say
```
p1 = column(4) + column(5) > 0 
p2 = if column(4) + column(5) > 0 then 1.0 / column(4) > 0 else false
```
then I do not think it is appropriate to reduce this to 
```
p1 = column(4) + column(5) > 0 
p2 = 1.0 / column(4) > 0
```
The second predicate which was previously guarded and would not error can now error. The order of the predicates might change (e.g. `p2` could be pushed down, as it depends on fewer columns) and we have no introduced an erroring behavior where one did not exist before.

wangandi commented:
> I'm not sure that this optimization is correct, and it would be great to discuss to determine if it is. If I have two predicates, let's say
> 
> ```
> p1 = column(4) + column(5) > 0 
> p2 = if column(4) + column(5) > 0 then 1.0 / column(4) > 0 else false
> ```
> 
> then I do not think it is appropriate to reduce this to
> 
> ```
> p1 = column(4) + column(5) > 0 
> p2 = 1.0 / column(4) > 0
> ```
> 
> The second predicate which was previously guarded and would not error can now error. The order of the predicates might change (e.g. `p2` could be pushed down, as it depends on fewer columns) and we have now introduced an erroring behavior where one did not exist before.

That is a good point.  I considered that predicates can return true, false, or null, but not that they can return `EvalErrors`. That said, it's hard for me to determine correctness for subexpressions that can return errors because I don't know what we guarantee about predicate order and returning errors. 

Based on what [SQL Server](https://techcommunity.microsoft.com/t5/sql-server/predicate-ordering-is-not-guaranteed/ba-p/383075) and [Oracle](https://asktom.oracle.com/pls/apex/f?p=100:11:0::::P11_QUESTION_ID:696647495510) do, I am in favor of us only having the guarantee:
"If there is a `CASE` statement, then we only evaluate the `THEN` statement if the `CASE` part evaluates to `TRUE`. We only evaluate the `ELSE` statement if the `CASE` part evaluates to `FALSE` OR `NULL`."

If we only have this one guarantee, then this PR should maintain correctness as long as we never substitute anything inside of `MirScalarExpr::If::cond` (and I can change the PR accordingly).

frankmcsherry commented:
> If we only have this one guarantee, then this PR should maintain correctness as long as we never substitute anything inside of MirScalarExpr::If::cond (and I can change the PR accordingly).

Right, I think this is the same resolution `column_knowledge.rs` had, and seems probably appropriate. Elsewhere in the `transform` crate we take the position that we may arbitrarily re-order predicates in particular (and perhaps expressions too).

frankmcsherry commented:
I think reworking the optimization to have any specific behavior on `mz_sleep` probably isn't the right way to handle the issue with it. If nothing else I definitely don't think we should add an undocumented `nondeterministic()` method to `UnaryFunc` without spelling out what this means.

wangandi commented:
Reverted the `mz_sleep` handling commit and instead created an issue about `mz_sleep` handling at #6262. Any other concerns with the PR?

frankmcsherry commented:
The error seems to be a probably minor or-normalization change. Not sure why but doesn't seem hard to fix (idk if the code or test is in error).

wangandi commented:
> The error seems to be a probably minor or-normalization change. Not sure why but doesn't seem hard to fix (idk if the code or test is in error).

The error is because I rebased this branch on main after merging the other PR and forgot about the fact that the other PR normalizes "and/or" statements, so the expected results of tests on this branch also need to be normalized now.

frankmcsherry commented on src/expr/src/relation/canonicalize.rs:
```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+    // Seed `predicates_to_apply` with the `Vec` of all predicates.
+    std::mem::swap(&mut predicates_to_apply, predicates);
+
+    while let Some(predicate_to_apply) = predicates_to_apply.pop() {
+        let mut replace_subexpr_other_predicates =
+            |expr: &MirScalarExpr, constant_bool: &MirScalarExpr| {
+                // Do not replace subexpressions equal to `expr` if `expr` is a
```
I think I understand this now, but the structure seems pretty convoluted. This is (correct me if wrong) the "worklist" pattern where things in the worklist are todo, and in the course of working things are dequeued, processed, and may involve putting things back in the queue. If that's right, I think this description (or something akin to it) is more helpful than the mechanical description up above (i.e. "take this out, put this in").

Also, it's pretty hard to grok with the closures that capture unclear state and .. idk whose names aren't quite enough to understand what they do. Like, I had to go read the implementation to determine that `visit_mut_replace_subexpr`'s main feature is that it doesn't process the `cond` block in `If` expressions.

Anyhow, I think this makes sense to me now, but I think each reader who comes to this will need to read all of the code to understand clearly what is happening. If you can add some comments at choice locations that reduces that property I'd be delighted!

frankmcsherry commented on src/expr/src/relation/canonicalize.rs:
```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+                }
+            };
+
+        if let MirScalarExpr::CallUnary {
+            func: UnaryFunc::Not,
+            expr,
+        } = &predicate_to_apply
```
For my style preferences, this would be a great place for a comment to call out that this is the meat of the method. Somehow having read through everything above, I didn't realize that this is where the behavior for this method actually starts.

frankmcsherry commented on src/expr/src/relation/canonicalize.rs:
```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+    false
+}
+
+fn visit_mut_replace_subexpr<F>(e: &mut MirScalarExpr, f: &mut F, changed: &mut bool)
+where
+    F: FnMut(&mut MirScalarExpr) -> bool,
```
Some doc comments on this would really help! Or perhaps `replace_subexprs_not_cond` as a name? Would this be valuable in `MirScalarExpr`? I know there is very similar logic in `column_knowledge.rs`, and it might make sense to promote to "a thing one shouldn't try and do by hand". No need to do it in this PR, but sounding out the idea to see if it makes sense.

wangandi replied:
I want to create a common method in `MirScalarExpr` for this because I know of similar code in the `JoinInputMapper`, but 1) I was afraid of the common method creating confusion because there would be changes to completely unrelated parts of the code. 2) I suspected that there were other parts of the code with similar logic that I did not know about, and I didn't want to spec out a common method until I saw those other parts. If you know of other places besides `column_knowledge.rs` and `join_input_mapper.rs` with similar logic, please let me know!

frankmcsherry reviewed (approved):
I left a pile of comments. Nothing is blocking, but there was at least one point where I wasn't clear on a limitation (the "cannot be null" requirement) and a few opportunities for leaving the code in a state that the next reader can more easily approach it, which is easier said than done. I'll leave those calls up to you!

wangandi commented on src/expr/src/relation/canonicalize.rs:
```diff
@@ -83,6 +87,123 @@ pub fn canonicalize_predicates(predicates: &mut Vec<MirScalarExpr>, input_type:
+    // `true` (resp. `false`) every subexpression in
+    // `(p1 & ... & p(i-1) & p(i+1) ... & pn)` that is equal to `pi`.
+
+    // (As for the reduction does not work if `p` can be null: if `p` is null,
+    // then p & f(p) becomes `null & f(null)`. `null & f(null)` is not
+    // necessarily equal to `null & f(true)` because `null & true = null`
+    // and `null & false = false`.)
```
> (we are in the special circumstance of evaluating predicates, rather than just performing general expression simplifications)

That is a good point. I was hesitant about collapsing the `null` and `false` answers, but now that I think about it, we don't have any transforms that would propagate information in such a way that Filter `p&q` turns into something like Filter `!(p&q)`.

frankmcsherry replied:
I could totally be wrong, too! :D
